### ISSUE TITLE
taxonomic rank browser

### ISSUE THREAD

**Issue by touretzkyds** [touretzkyds/KnowledgeGraphDemo#59]
This is a specialized view for exploring taxonomic rank information, and an opportunity for us to try out new visualization tools. If the user is viewing the anmal taxonomy and they click on a green Genus node, we want to open a new panel with this taxonomic rank browser view.

The view will display the taxonomic ranks from Species up to Kingdom as a vertical chain of nodes connected by nextHighestRank and nextLowestRank nodes. We'll zoom in on Genus (because that's what the user clicked on) and display its properties. The navigation arrow will allow us to move up or down the chain.

What's interesting is what happens if the user clicks on some attribute like "taxonomy" or "third-order class". We want to show that subgraph in the same view, off to the side of the rank chain. So now we have two subgraphs: the rank chain, and the subgraph for the new concept we clicked on, which is a property of the rank we were focused on.

Let's use cola.js to do the graph layout for this viewer. It's a constraint-based layout framework, so we can force the taxonomic ranks into a vertical chain by specifying vertical distance constraints on the rank nodes with horizontal distance set to 0. Similarly, we can cause the "taxonomy" subgraph to be off to the side of the main chain by placing a horizontal distance constraint. See this cola.js interactive demo for an example of spawning new subgraphs from a main graph (click on an image to expand the node): https://ialab.it.monash.edu/webcola/examples/browsemovies.html

cola.js is designed to lay out small graphs. We don't want to use it for the animal taxonomy because that tree is huge, meaning the number of constraints the system would have to deal with would be very large. Also, we just want a simple tree structure for the animal taxonomy so we don't need fancy constraint satisfaction algorithms to incrementally "solve" for a layout; we can just compute it directly. But if we decide we like cola.js then @nhadlaw might consider using it in the taxon property view in place of our current property layout algorithm.

**Comment by deyaliao** [comment]
Hi, was just wondering if there was a way to return the taxonomic ranks in the order of their hierarchy? My query returns all the taxonomic ranks, but not in the correct order.
```
PREFIX kgo: <http://solid.boltz.cs.cmu.edu:3030/ontology/>
          PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
          PREFIX skos: <http://www.w3.org/2004/02/skos/core#>
          PREFIX boltz: <http://solid.boltz.cs.cmu.edu:3030/data/> 

          SELECT ?rank ?rankLabel ?definition ?nextHigher ?nextLower WHERE {
            ?rank kgo:instanceOf boltz:Q427626 ;
                  rdfs:label ?rankLabel ;
                  skos:definition ?definition .
  OPTIONAL{?rank kgo:nextHigherRank ?nextHigher} .
  OPTIONAL{?rank kgo:nextLowerRank ?nextLower}.
          } 
```
This makes creating the vertical constraint for all ranks slightly more difficult, as it's easier to arrange the nodes of the query results come in the right order. Of course, this shouldn't be the only way to do it, but I was just wondering if there was indeed a way.

**Comment by touretzkyds** [comment]
You can order the ranks by distance from the highest rank. This StackOverflow post shows how to measure that using property paths and the COUNT function: https://stackoverflow.com/questions/18024413/finding-all-steps-in-property-path

While you can probably make this work, it's kind of hairy. You might be better off just retrieving a list of ranks and their next higher rank, and then sorting the list. Here's a simple sort algorithm:

Assume you have a bunch of pairs (p,q) where p and q are taxon ranks and you have triples (p nextHigherRank q). Then do the following:
````
nodes = {}
done = False
WHILE ! done:
  done = True
  FOR (p,q) in pairs:
    IF p not in nodes:
      nodes[p] = 0
      nodes[q] = 0
    IF nodes[q] < nodes[p] + 1
      nodes[q] = nodes[p] + 1
      done = False
````
We keep going until the value of nodes stops changing.

**Comment by deyaliao** [comment]
Hi, I was trying to remember if we reached a conclusion about the type/isInstanceOf repetition of concepts? Specifically:
<img width="233" alt="Screen Shot 2022-07-25 at 11 28 42 AM" src="https://user-images.githubusercontent.com/61635230/180848624-1bb1d367-0f83-4816-bbef-62fc42a269df.png">

**Comment by touretzkyds** [comment]
I think we decided that rdf:type should point to rdfs:Class, and kgo:instanceOf should point to the "taxonomic rank" concept as it does now.

**Comment by deyaliao** [comment]
Yes, I changed it as so. Another concern: I'm trying to change the source code for expand-collapse to see if there's a way to stop the creation of an expandable node if my data has an additional property that describes the object as non expandable. I've looked through the entire code, have added many console.logs at various points I think the code would turn in nodes into expandable ones, but I'm still not able to figure it out. [Here](https://github.com/iVis-at-Bilkent/cytoscape.js-expand-collapse/blob/master/cytoscape-expand-collapse.js) is the source code. 

Another way would be to on my end, remove the expandable property of the node if I find that its data indicates it's not supposed to be expandable. However, nothing in the [documentation](https://github.com/iVis-at-Bilkent/cytoscape.js-expand-collapse) indicates that this is possible. 

Of course, I could just make each individual nested compound group all expandable/collapsable nodes. This means all nodes (with additional properties to expand) would have the plus node on the top left corner (as shown in the screenshot below) and would be expanded as such. This is a fine approach too, but my code would just have to be slightly different. 
<img width="482" alt="Screen Shot 2022-07-27 at 6 06 47 PM" src="https://user-images.githubusercontent.com/61635230/181398476-7eb57ce2-3eea-4e7a-bef5-84c65263ba4b.png">

**Comment by touretzkyds** [comment]
I took a quick look at the code and found a function called drawExpandCollapseCue.  Did you try modifying that? What happened?

I don't think we want expand/collapse icons on every node; that would be ugly.

I need to give another progress report at 1pm on Friday. Can you check in whatever you have today, even if it's not fully working, so I can play with it and show something to my colleagues on Friday? Thanks.

**Comment by deyaliao** [comment]
I tried modifying that, but that is solely in charge in adding the node icon and making it clickable. It rendered the node not able to expand (which is what we want), but the internal data was already changed to mirror that of an expand-collapse node. I tried adding a property in the internal data called ("expandable": true/false). Usually, if you call cy_graph.$id(right_id).data(), it returns all of the properties about the node, including this "expandable" one. However, whenever the node turns into an expand-collapse node, it adds a bunch of new properties, and for some reason removes this property about the "expandable". So, I have no way to truly access whether if it was an expandable node or not. Basically, I can brute force it to always not be able to expand, but there is no way to ensure that the node is supposed to be (in our terms) expandable. I am trying to find the area where the expand-collapse extension turns the node into a expandable-collapsible node precisely.

I pushed something to demo2e. I was playing around with the expand collapse the entirety of today, but I don't think it's at a state where one can comfortably play around with it, so I pushed the version I had before. All green nodes are those with Q-values, and those should all be clickable to display its properties. When you click on those nodes, make sure to drag them out a bit so when its properties display, it doesn't get in the way of the existing nodes. I haven't implemented something to automate this yet.

**Comment by touretzkyds** [comment]
This version of demo2e looks pretty bad: all the taxonomic ranks are at the same coordinates, so they're on top of each other. Can you put this back to its previously less broken state?

**Comment by touretzkyds** [comment]
Looking more closely at cytoscape-expand-collapse.js, I see that line 1044 has a call node.addClass('cy-expand-collapsed-node') and lines 740 and 2081 have calls to removeClass('cy-expand-collapsed-node').  Could these have something to do with the loss of the internal data item you added? How exactly are you adding this item? There's also a call to restore() at line 744; I wonder what that's doing.

**Comment by deyaliao** [comment]
Hi, I’m not sure what you mean by all the taxonomic ranks are the same coordinates. Could you send a screenshot, since the version I pushed is a more advanced version compared to the one from last meeting?

**Comment by touretzkyds** [comment]
Here is what I'm seeing right now. I did a git pull to make sure I was up to date.

![deya1](https://user-images.githubusercontent.com/1978092/181674702-171a18bf-bb10-4ed7-87e8-2ec6ad483534.png)
![deya2](https://user-images.githubusercontent.com/1978092/181674789-966547c2-d340-428b-b0b7-511817d5d457.png)

**Comment by deyaliao** [comment]
I'm not sure what's going on, but I do remember this happening before in previous meetings––changes that didn't reflect on your end, despite the github being updated? My Demo2e looks like. I tried the link of your demo, and it does look like it's not updated. I pushed again just in case (but no changes were made, so everything was up to date). 
<img width="815" alt="Screen Shot 2022-07-28 at 9 28 29 PM" src="https://user-images.githubusercontent.com/61635230/181683677-2d169013-9ad1-4477-b90d-308565f21acb.png">

**Comment by touretzkyds** [comment]
Yes, the problem is the same as last time: you added a new file and forgot to add it to the github repository. We're missing cytoscape-cola.js.

**Comment by touretzkyds** [comment]
If you do "git status" it will tell you if you have files in the directory that are not in the repository.

**Comment by deyaliao** [comment]
Sorry, I see; try again now? Thanks.

**Comment by touretzkyds** [comment]
Success! Just what I need for my demo tomorrow. Thanks.

**Comment by deyaliao** [comment]
I'll be more careful next time about the files.

**Comment by deyaliao** [comment]
I actually found a way to increase the distance between nearby edges. 
<img width="269" alt="Screen Shot 2022-08-01 at 10 21 44 PM" src="https://user-images.githubusercontent.com/61635230/182297990-c2576329-3b44-42e8-aa7e-df356fb0b01a.png">
Do you want something like this, or perhaps make the vertical edge push out the other way––so they're both curved?

**Comment by touretzkyds** [comment]
This is exactly what I was hoping for. Perfect!

**Comment by deyaliao** [comment]
I had a question about an interaction on our end. Say the user starts off here, and they want to click the "taxonomy" node, which is a _facetOf_ this node. 
<img width="542" alt="Screen Shot 2022-08-09 at 4 36 52 PM" src="https://user-images.githubusercontent.com/61635230/183779670-ff17d4d3-1798-485c-83c7-03daa570aa48.png">

This displays the properties of that node, which is shown below. As you see in the top left corner, there is a _hasQuality_ arrow that points to a "taxonomic rank" node, which is the same node that we started off with in the first image.
<img width="646" alt="Screen Shot 2022-08-09 at 4 39 39 PM" src="https://user-images.githubusercontent.com/61635230/183779953-807951ab-1887-4821-aff9-423320e62d81.png">
 
Basically, these nodes are practically inverses of each other. When I then click on the "taxonomic rank" node in the top left corner again, do I a) open up a new tab with the same properties displayed in the first image, or 2) create new edges that point to existing properties (after all, they are the same?). There are issues with both:

1) This would lead to an endless expansion of subpanels that are identical to each other. Say a panel for "taxonomic rank" is displayed again; if I click on "taxonomy" node, it would create the same panel that already exists. This seems like an endless loop.
2) If two nodes with this inverse property have many nodes in between them (say 4-5), it would be very cluttered to create these edges that connect the node with its properties many nodes apart. Visually, this would be a pain.

Right now, when I click on the node, it just doesn't do anything visually, but turns it into an expand-collapse node. 
<img width="531" alt="Screen Shot 2022-08-09 at 5 01 31 PM" src="https://user-images.githubusercontent.com/61635230/183782063-d295c185-6138-4b02-8cf4-cbe22b46e332.png">

Thanks.

**Comment by touretzkyds** [comment]
Yes, this is a problem, and there probably isn't a single right approach to dealing with it. What  you're doing right now, just turning the redundant node into an expand-collapse node, is reasonable. Here are a couple of other ideas. First, let's define some terminology: we'll refer to the original "taxonomic rank" node as node A and the facetOf property as P1. We'll refer to the "taxonomy" node as node B and the hasQuality property as P2. We'll refer to the redundant "taxonomic rank" node as A'. Thus, your diagram above shows `A  =P1=>  B  =P2=>  A'`.

1. In cases like the above where we have just one node B between A and A', we can decline to create a separate A' node and instead make the P2 arrow point back to the original A node.
2. In cases where there are multiple nodes between A and A', we can make each subpanel a different color, and we can replace A' with a special shape and a color that matches the subpanel we're referencing. If the user clicks on the special shape, instead of expanding that node again we simply bring the associated subpanel into view if it's off-screen or covered by something.

**Comment by touretzkyds** [comment]
demo2e is broken again. This time it's because cy.automove is undefined.

**Comment by deyaliao** [comment]
Oh, this was an extension that I initially imported but never used. I thought I deleted where I initialized it, but I guess not. I just changed it (with some new features).
